### PR TITLE
Optimize ArcContainer.getContexts() used during client proxy invocation

### PR DESCRIPTION
**Pull request by mkouba**
This pull request supersedes https://github.com/quarkusio/quarkus/pull/27804.

### CODE REVIEW

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ArcContainerImpl.java [hunk 1]
```diff
@@ -149,6 +153,10 @@ public ArcContainerImpl(CurrentContextFactory currentContextFactory) {
             qualifierNonbindingMembers.putAll(components.getQualifierNonbindingMembers());
         }
 
+        List<InjectableContext> requestContexts = contexts.get(RequestScoped.class);
+        this.requestContextSingleton = requestContexts.size() == 1 ? Collections.singletonList(requestContexts.get(0)) : null;
```
Comment (Sanne) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ArcContainerImpl.java:157]: Why would you store `null` when you could store the list?

Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ArcContainerImpl.java:157]: Sorry, that was an old version...

**Comment by mkouba** [comment]
@Sanne I've added the `Contexts` abstraction which is built on top of your idea of `ContextsMap`.

**Comment by mkouba** [comment]
I'll run the benchmarks tomorrow...

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                    for (int i = 0; i < mapSize; i++) {
+                        if (keys[i].equals(type)) {
+                            return values[i];
+                        }
+                    }
+                    return Collections.emptyList();
+                }
+            };
+            this.size = 3 + mapSize;
+            Set<Class<? extends Annotation>> all = new HashSet<>(contexts.keySet());
```
Comment (Ladicek) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java:78]: If `contexts` are not empty, but do _not_ contain a custom implementation of the request context, then `all` won't include `RequestScoped.class` I think?

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                protected List<InjectableContext> computeValue(Class<?> type) {
+                    for (int i = 0; i < mapSize; i++) {
+                        if (keys[i].equals(type)) {
+                            return values[i];
+                        }
+                    }
+                    return Collections.emptyList();
+                }
+            };
+            this.size = 3 + mapSize;
```
Comment (Ladicek) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java:77]: This looks to me like `mapSize` is always 1 more than necessary (and the last element of `keys` and `values` is always `null`), and `size` is only accurate if there's no custom implementation of the request context.

IMHO, `mapSize` should be just `contexts.size()`, and `size` should be `2 + (contexts.containsKey(RequestScoped.class) ? 0 : 1) + mapSize`.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            }
+            this.unoptimizedContexts = new ClassValue<>() {
+                @Override
+                protected List<InjectableContext> computeValue(Class<?> type) {
+                    for (int i = 0; i < mapSize; i++) {
+                        if (keys[i].equals(type)) {
+                            return values[i];
+                        }
+                    }
+                    return Collections.emptyList();
```
Comment (Sanne) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java:74]: are you sure you want to associate an emptyList to any type which is not in the map?
In my implementation I made sure to differentiate, so I could return `null` - but perhaps I was going too far, as I just wanted to emulate exactly the same semantics as you had before.

So perhaps it's fine and perhaps it doesn't matter, but wanted to call out the semantic difference.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java (hunk 2 above)
Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java:78]: You're right, that's an oversight... it was there a few iterations earlier.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java (hunk 4 above)
Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java:74]: `ArcContainer.getContexts(Class<? extends Annotation>)` never returns `null` so I think that an empty list is ok.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java (hunk 3 above)
Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java:77]: Yep, an oversight again..

Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java:77]: Thanks for noticing! I will add some tests...

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+    Contexts(ManagedContext requestContext, InjectableContext applicationContext,
+            InjectableContext singletonContext, Map<Class<? extends Annotation>, List<InjectableContext>> contexts) {
+        this.requestContext = requestContext;
+        this.applicationContext = applicationContext;
+        this.singletonContext = singletonContext;
+
+        this.applicationContextSingleton = Collections.singletonList(applicationContext);
+        this.singletonContextSingleton = Collections.singletonList(singletonContext);
+        // request singleton is only set if there is only one context registered for @RequestScoped (99% of cases)
```
Comment (Ladicek) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java:48]: This comment is not true? It seems to me that `requestContextSingleton` is always set to a non-`null` value. (And that seems correct, from what I can tell. The comment is probably just stale?)

**Review by Ladicek — APPROVED** [review]
LGTM, modulo one stale comment :-)

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java (hunk 5 above)
Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java]: Yes, it is :facepalm:

**Comment by mkouba** [comment]
FTR the [`RequestScopedProxyInvocationBenchmark`](https://github.com/mkouba/arc-benchmarks/blob/master/src/main/java/io/quarkus/arc/benchmarks/RequestScopedProxyInvocationBenchmark.java) shows ~32% improvement.

**Pull request merged**